clang/lib/CIR/CodeGen/CIRGenBuiltinAArch64.cpp

```diff
@@ -2261,6 +2261,37 @@ mlir::Value emitNeonCall(CIRGenBuilderTy &builder,
       .getResult();
 }
 
+template <typename Operation>
+static mlir::Value
+emitNeonCall(CIRGenBuilderTy &builder, llvm::SmallVector<mlir::Type> argTypes,
```

Comment:
As part of this refactoring, you should also change the non-templated version of `emitNeonCall` to call into this so that any future changes hit both paths at once.

Reply:
I think I cannot do that. Because the non-templated version of `emitNeonCall` doesn't know the stringref name maps to which operation. However, I can do something to reduce code duplication. I extract `prepareNeonCallArgs`. I leave ConstrainedFPCall handling behand. We can make it better after removing this missing feature.

Reply:
Why can't the non-template just do (as a fictional example):
```
return emitNeonCallToOp<cir::LLVM_CallIntrinsicOp>(builder, {arg0.getType()}, args, "",
                                          getCIRGenModule().FloatTy,
                                          getLoc(E->getExprLoc()));
```
Please also rename the templated version to `emitNeonCallToOp`